packages/web/src/server.tsx

```diff
@@ -53,32 +56,79 @@ function getScripts(statsStr: string) {
     }, assets);
 }
 
-app.get('*', (req, res) => {
-  const html = ReactDOMServer.renderToString(
+function renderPass(store: AppStore, location: string) {
+  return ReactDOMServer.renderToString(
     <Provider store={store}>
-      <StaticRouter location={req.url}>
+      <StaticRouter location={location}>
         <App locationChangeHandler={() => {}} />
       </StaticRouter>
     </Provider>
   );
+}
+
+async function renderTimeout(timeout: number, abort: AbortController) {
+  await setTimeout(timeout, null, { signal: abort.signal });
+  throw new Error('Timed out');
+}
+
+async function renderCycle(location: string, maxPasses = 3, passTimeout = 15000) {
```

Comment:
How are default values determined? I mean why it's 3 passes and not 2 or 5

Reply:
Determined by the requirements of the application. Effectively, this is a number longest chain of dependent requests (waterfall requests) required to render an application. Given that we've no dependent requests at the moment, two passes would be sufficient. I'll keep track and update this number as needed.